pkg/s3select/input.go

```
@@ -0,0 +1,322 @@
+// as if the record were empty.
+//
+// In general, this is a very tolerant of problems  reader.
+func (reader *Input) ReadRecord() []string {
+	var row []string
+	var fileErr error
+
+	if reader.firstRow != nil {
+		row = reader.firstRow
+		reader.firstRow = nil
+		return row
+	}
+
+	row, fileErr = reader.reader.Read()
+	emptysToAppend := reader.minOutputLength - len(row)
+	if fileErr == io.EOF {
+		return nil
+	} else if _, ok := fileErr.(*csv.ParseError); ok {
+		emptysToAppend = reader.minOutputLength
+	}
+
+	if emptysToAppend > 0 {
+		for counter := 0; counter < emptysToAppend; counter++ {
+			row = append(row, "")
+		}
+	}
+
+	return row
+}
+
+// convertMySQL Replaces double quote escape for column names with backtick for
+// the MySQL parser
+func convertMySQL(random string) string {
+	return strings.Replace(random, "\"", "`", len(random))
+}
+
+// readHeader reads the header into the header variable if the header is present
+// as the first row of the csv
+func (reader *Input) readHeader() error {
+	var readErr error
+	if reader.options.HasHeader {
+		reader.firstRow, readErr = reader.reader.Read()
+		if readErr != nil {
+			return ErrCSVParsingError
+		}
+		reader.header = reader.firstRow
+		reader.firstRow = nil
+		reader.minOutputLength = len(reader.header)
+	}
+	return nil
+}
+
+// createStatXML is the function which does the marshaling from the stat
+// structs into XML so that the progress and stat message can be sent
+func (reader *Input) createStatXML() string {
+	if reader.options.CompressedGzip {
+		reader.stats.BytesProcessed += processSize(reader.header)
+	}
+	if !reader.options.CompressedGzip {
+		reader.stats.BytesProcessed = reader.options.StreamSize
+		reader.stats.BytesScanned = reader.stats.BytesProcessed
+	}
+	statXML := &stats{
+		BytesScanned:   reader.stats.BytesScanned,
+		BytesProcessed: reader.stats.BytesProcessed,
+		BytesReturned:  reader.stats.BytesReturned,
+	}
+	out, _ := xml.MarshalIndent(statXML, "", "   ")
```

Don't ignore the error here we can send an error back to the client here, also do we need to Indent?